### PR TITLE
AWS Lambda Docs 

### PR DESCRIPTION
**Pull request by aryans1204**
Added docs for AWS Lambda

### CODE REVIEW

File: docs/aws/aws-lambda.mdx [hunk 1]
```diff
@@ -3,3 +3,133 @@ sidebar_position: 5
...
+
+Let's dive into what AWS Lambda is, how you can use it, why you should use it, and some important points to remember while using it.
+
+#  What is AWS Lambda?
+
+In recent years, serverless computing paradigms have become increasingly popular for organizations and companies to develop and deploy their applications.
+But, what is serverlss computing? Does there mean there are no servers involved? Well, serverless computing still has servers to host application data for deployment,
+it abstracts away the need to manage servers for application developers. How it achieves this is by providing FaaS(Function as a Service) platform whereby 
+app developers can simply write cloud native functions in a higher level general purpose language, like Python or Go, to automate tasks of autoscaling or software
+patches. As opposed to IaaS(Infrastructure as a Service) models, where users must prepurchase units of compute like VM's and perform autoscaling and load balancing
```
Comment (chanbakjsd) [inline docs/aws/aws-lambda.mdx:15]:
```suggestion
patches. As opposed to IaaS (Infrastructure as a Service) models, where users must prepurchase units of compute like VM's and perform autoscaling and load balancing
```

File: docs/aws/aws-lambda.mdx [hunk 2]
```diff
@@ -3,3 +3,133 @@ sidebar_position: 5
...
+themselves, the FaaS model provides an event-driven execution model, where users need not prepurchase or manually scale any units of compute, but rather 
+autoscaling is performed based on needs and the user only pays for what they use. Most cloud providers have some form of FaaS service, and Lambda is AWS's FaaS service.
+
+With AWS Lambda, the apps are only launched as needed, as opposed to running a EC2 instance persistently. This provides some serious advantages, one of them being
+that the app is only launched when an event(API request, or a client request) triggers the app to run, and the cloud provider will dynamically handle the resource
```
Comment (chanbakjsd) [inline docs/aws/aws-lambda.mdx:20]:
```suggestion
that the app is only launched when an event (API request, or a client request) triggers the app to run, and the cloud provider will dynamically handle the resource
```

File: docs/aws/aws-lambda.mdx [hunk 3]
```diff
@@ -3,3 +3,133 @@ sidebar_position: 5
...
+allocation for the code to serve all requests efficiently. The app will stop running once the requests have been served, and the user will only be charged for the
+duration of runtime. Not only does AWS Lambda have clear efficiency and cost benefits, it also abstracts away the processes of operating system maintenance,
+security patching, logging, server provisioning, load balancing to the cloud provider, so YOU, the app developer can simply focus on developing your app.
+
+#  How does AWS Lambda work?
+
+Now that we've takena look at some functionalities offered by AWS Lambda, how exactly does it work, and how can you, as an app or API developer leverage the
```
Comment (chanbakjsd) [inline docs/aws/aws-lambda.mdx:27]:
```suggestion
Now that we've taken a look at some functionalities offered by AWS Lambda, how exactly does it work, and how can you, as an app or API developer leverage the 
```

File: docs/aws/aws-lambda.mdx [hunk 4]
```diff
@@ -3,3 +3,133 @@ sidebar_position: 5
...
+With AWS Lambda, the apps are only launched as needed, as opposed to running a EC2 instance persistently. This provides some serious advantages, one of them being
+that the app is only launched when an event(API request, or a client request) triggers the app to run, and the cloud provider will dynamically handle the resource
+allocation for the code to serve all requests efficiently. The app will stop running once the requests have been served, and the user will only be charged for the
+duration of runtime. Not only does AWS Lambda have clear efficiency and cost benefits, it also abstracts away the processes of operating system maintenance,
+security patching, logging, server provisioning, load balancing to the cloud provider, so YOU, the app developer can simply focus on developing your app.
+
+#  How does AWS Lambda work?
+
+Now that we've takena look at some functionalities offered by AWS Lambda, how exactly does it work, and how can you, as an app or API developer leverage the 
+functionlities of AWS Lambda?
```
Comment (chanbakjsd) [inline docs/aws/aws-lambda.mdx:28]:
```suggestion
functionalities of AWS Lambda?
```

File: docs/aws/aws-lambda.mdx [hunk 5]
```diff
@@ -3,3 +3,133 @@ sidebar_position: 5
...
+
+As mentioned before, AWS Lambda is a event-driven execution model. Specifically, the user would write Lambda functions in a higher level programming language,
+like Python or Go which achieve the desired logic. These functions would then be run on lightweight containers whenever an event triggers these functions to run.
+The containers are:
+
+* Stateless- any changes made to the container during execution are not saved after container is shut down. They do not store state.
+* Ephemeral- they run for a very short time during execution and do not store any state
+* Event-Triggered- they run only when the associated event triggers them(we'll look at this in depth later in the doc)
+* Fully managed- they are managed by the cloud provider, which means the app developer does not have to create Container Images or monitor the health of the containers.
```
Comment (chanbakjsd) [inline docs/aws/aws-lambda.mdx:37]:
```suggestion
* Stateless - any changes made to the container during execution are not saved after container is shut down. They do not store state.
* Ephemeral - they run for a very short time during execution and do not store any state
* Event-Triggered - they run only when the associated event triggers them(we'll look at this in depth later in the doc)
* Fully managed - they are managed by the cloud provider, which means the app developer does not have to create Container Images or monitor the health of the containers.
```

File: docs/aws/aws-lambda.mdx [hunk 6]
```diff
@@ -3,3 +3,133 @@ sidebar_position: 5
...
+
+Let's take a look at a hands on example at how an app developer can write Lambda functions. The example here is a Lambda function which is triggered at 11:59 PM
+SGT every night, to check for any running EC2 instances and automatically shut them down, to reduce unnecessary costs.
+
+The process of using AWS Lambda can be decoupled into two parts:
+
+#   Writing the Lambda function
```
Comment (chanbakjsd) [inline docs/aws/aws-lambda.mdx:44]:
```suggestion
# Writing the Lambda function
```

File: docs/aws/aws-lambda.mdx [hunk 7]
```diff
@@ -3,3 +3,133 @@ sidebar_position: 5
...
+
+AWS Lambda provides support for most general purpose programming languages, from Python, Go, Java, JavaScript and Ruby. As mentioned earlier, the Lambda function 
+is an actual function, that represents the logic of the task you want to peform, essentially, in the function body you will code what you want the Lambda function
```
Comment (chanbakjsd) [inline docs/aws/aws-lambda.mdx:47]:
```suggestion
is an actual function, that represents the logic of the task you want to perform, essentially, in the function body you will code what you want the Lambda function
```

File: docs/aws/aws-lambda.mdx [hunk 8]
```diff
@@ -3,3 +3,133 @@ sidebar_position: 5
...
+   status = client.describe_instance_status(IncludeAllInstances=True)        
+   def lambda_handler(event, context):                                      
+     for i in status["InstanceStatuses"]:                                    
+         if i["InstanceStatus"] == "running":                                
+             client.stop_instances(InstanceIds=i["InstanceId"])
+```              
+
+
+
+The function can be given any name, in this case we just call it event_handler and define it just like any other function. Each function must take as argument
```
Comment (chanbakjsd) [inline docs/aws/aws-lambda.mdx:71]:
```suggestion
The function can be given any name, in this case we just call it `lambda_handler` and define it just like any other function. Each function must take as argument
```

File: docs/aws/aws-lambda.mdx [hunk 9]
```diff
@@ -3,3 +3,133 @@ sidebar_position: 5
...
+event and context, which represent the particular event that triggers the Lambda function and the context under which the lambda function is triggered. While it 
+is not necessary to use thee arguments in the fucntion body, nonetheless, every Lambda function must have these as function arguments. A Lambda function may 
+have more arguments too, dpeending upon the nature of the event that triggers it. The logic of the function is pretty simple. We perform a brute force search of all
+EC2 instances using boto3, an AWS SDK for Python, and of those that are running, we stop those instances. We'll save this as a Python file.
+
+# .  Deploying the Lambda function
```
Comment (chanbakjsd) [inline docs/aws/aws-lambda.mdx:77]:
```suggestion
# Deploying the Lambda function
```

File: docs/aws/aws-lambda.mdx [hunk 10]
```diff
@@ -3,3 +3,133 @@ sidebar_position: 5
...
+2. Go to the AWS Lambda page by searching on the search bar.
+3. Click on Create Function on the top right
+4. You'll see multiple options to deploy the function. This also includes deploying a container image or choosing from available Lambda function repos. While
+these options might be relevant during API or app deployment, for our use case we will simply choose Author from Scratch.
+5. Name your Lambda function. Note that it is not necessary to name the Lambda function the same as the underlying function, it is best practice to name a Lambda
+function in a manner that clearly describes its purpose.
+6. Choose Python as your runtime, since this is a Python function.
+7. Choose the default IAM Role, however, for other Lambda functions the user can configure new IAM Roles or choose existing ones to limit the permissions and 
+access of the Lambda function.
+8. Now that your Lambda function is created, we need to add an EVent rule which dictates when the Lambda function will be triggered. Click on your Lamda function
```
Comment (chanbakjsd) [inline docs/aws/aws-lambda.mdx:91]:
```suggestion
8. Now that your Lambda function is created, we need to add an event rule which dictates when the Lambda function will be triggered. Click on your Lambda function
```

File: docs/aws/aws-lambda.mdx [hunk 11]
```diff
@@ -3,3 +3,133 @@ sidebar_position: 5
...
+and click on Test to test your function. If everything goes fine, you should be status Successful. Additionally if you want, you could actually try running EC2 
+instances and Testing the Lambda function top see if the function actually stops the instances or not. It is a best practice to thoroughly test your Lambda
```
Comment (chanbakjsd) [inline docs/aws/aws-lambda.mdx:93]:
```suggestion
instances and Testing the Lambda function above to see if the function actually stops the instances. It is a best practice to thoroughly test your Lambda
```

File: docs/aws/aws-lambda.mdx [hunk 12]
```diff
@@ -3,3 +3,133 @@ sidebar_position: 5
...
+functions on all test cases before deploying them.
+9. On the top left corner, you will find a button for Add Trigger. Click on that, and select EventBridge as the trigger. AWS EventBridge is another service
+that monitors events at periodic times in the AWS console, which is relevant for us since ourt Lambda function is periodic too.
+10. Click on Create New rule. Enter a relevant rule name and description. This rule will define the schedule of our Lambda function.
+11. You'll find options between Event pattern or Schedule expression. Since our Lambda function is essentially a scheduled function to run at midnight everyday,
+go ahead and chooe Schedule expression.
```
Comment (chanbakjsd) [inline docs/aws/aws-lambda.mdx:99]:
```suggestion
go ahead and choose Schedule expression.
```

File: docs/aws/aws-lambda.mdx [hunk 13]
```diff
@@ -3,3 +3,133 @@ sidebar_position: 5
...
+12. Now we need to add a schedule expression which defines the schedule of the function. To do this, we'll use CRON, which is a Unix system utility used for
+automatic job scheduling. Enter this exact CRON expression to achieve the desired schedule "cron(59 11 * * ? 2022-2024)" without the quotes. For more information
+about writing Cron expressions in AWS, check out [AWS Cron Expressions](https://docs.aws.amazon.com/AmazonCloudWatch/latest/events/ScheduledEvents.html#CronExpressions)
+13. At this point, your AS Lambda function is complete, and ready to be deployed. Head back to the Lambda funtion page, and clikc on Deploy. Note: you may or may not
```
Comment (chanbakjsd) [inline docs/aws/aws-lambda.mdx:103]:
```suggestion
13. At this point, your AWS Lambda function is complete, and ready to be deployed. Head back to the Lambda function page, and click on Deploy. Note: you may or may not
```

File: docs/aws/aws-lambda.mdx [hunk 14]
```diff
@@ -3,3 +3,133 @@ sidebar_position: 5
...
+be able to actually deploy the LAmbda function based on the level of permissions provided to your IAM identity. Contact your System Administrator for more information.
```
Comment (chanbakjsd) [inline docs/aws/aws-lambda.mdx:104]:
```suggestion
be able to actually deploy the Lambda function based on the level of permissions provided to your IAM identity. Contact your System Administrator for more information.
```

File: docs/aws/aws-lambda.mdx [hunk 15]
```diff
@@ -3,3 +3,133 @@ sidebar_position: 5
...
+
+With this, we've seen the lifecycle of deploying an AWS Lambda function. While the function logic and event triggers of your functions may be very different, the
+general workflow is the same as described. Note that there are more options for setting Triggers other than EventBridge, like S3, Kafka, and API Gateway among many more
+that may be relevant to your use case.
+
+# . Why and When to use AWS Lambda?
```
Comment (chanbakjsd) [inline docs/aws/aws-lambda.mdx:110]:
```suggestion
# Why and When to use AWS Lambda?
```

File: docs/aws/aws-lambda.mdx [hunk 16]
```diff
@@ -3,3 +3,133 @@ sidebar_position: 5
...
+aggregate data by a defined logic based on an input data from Amazon S3, or other apps/databases.
+
+While AWS Lambda and serverless computing provide a ton of cool features, nonetheless, Lambda is not the answer for all your needs. For example, larger apps that require
+state and do not have a defined schedule of execution but feature wide ranging features and modules might be better served using microservices and container
+orchestration.
+
+
+
+
+
```
Comment (chanbakjsd) [inline docs/aws/aws-lambda.mdx:135]: Excessive whitespaces.

**Review by chanbakjsd — CHANGES_REQUESTED** [review]
Overall, this is a really comprehensive article on AWS Lambda.
Other than the trivial cosmetic changes I've added comments on, this article could use some shortening as it might be too verbose for someone to refer to easily.
Can you also replace the headings with subheadings (`##` or `###` instead of `#`)? That'll generate sections on the sidebar which will help navigation.

File: docs/aws/aws-lambda.mdx [hunk 17]
```diff
@@ -3,3 +3,142 @@ sidebar_position: 5
...
+* Ephemeral - they run for a very short time during execution and do not store any state
+* Event-Triggered - they run only when the associated event triggers them (we'll look at this in depth later in the doc)
+* Fully managed - they are managed by the cloud provider, which means the app developer does not have to create Container Images or monitor the health of the containers.
+
+Let's take a look at a hands on example at how an app developer can write Lambda functions. The example here is a Lambda function which is triggered at 11:59 PM
+SGT every night, to check for any running EC2 instances and automatically shut them down, to reduce unnecessary costs.
+
+The process of using AWS Lambda can be decoupled into two parts:
+
+##   Writing the Lambda function
```
Comment (chanbakjsd) [inline docs/aws/aws-lambda.mdx:46]:
```suggestion
## Writing the Lambda function
```

File: docs/aws/aws-lambda.mdx [hunk 18]
```diff
@@ -3,3 +3,142 @@ sidebar_position: 5
...
+
+
+
+The function can be given any name, in this case we just call it event_handler and define it just like any other function. Each function must take as argument
+event and context, which represent the particular event that triggers the Lambda function and the context under which the Lambda function is triggered. While it 
+is not necessary to use these arguments in the function body, nonetheless, every Lambda function must have these as function arguments. A Lambda function may 
+have more arguments too, depending upon the nature of the event that triggers it. The logic of the function is pretty simple. We perform a brute force search of all
+EC2 instances using boto3, an AWS SDK for Python, and of those that are running, we stop those instances. We'll save this as a Python file.
+
+##   Deploying the Lambda function
```
Comment (chanbakjsd) [inline docs/aws/aws-lambda.mdx:79]:
```suggestion
## Deploying the Lambda function
```

**Review by chanbakjsd — CHANGES_REQUESTED** [review]
In addition to the previous review, there are still some minor cosmetic issues that needs to be fixed.

File: docs/aws/aws-lambda.mdx [hunk 19]
```diff
@@ -3,3 +3,76 @@ sidebar_position: 5
 ---
 
 # AWS Lambda
+
+Let's dive into what AWS Lambda is, how you can use it, why you should use it, and some important points to remember while using it.
+The documentation is written from the persepctive of a reader who is completely new to serverless computing and AWS Lambda, and so there are some sections that may
```
Comment (chanbakjsd) [inline docs/aws/aws-lambda.mdx:8]:
```suggestion
The documentation is written from the perspective of a reader who is completely new to serverless computing and AWS Lambda, and so there are some sections that may
```

**Review by chanbakjsd — COMMENTED** [review]
Just a typographical error.
An idea to expand on the docs in the future is how to use AWS Lambda (common operations).

**Review by chanbakjsd — APPROVED** [review]
This is a good start for the document.

**Comment by aryans1204** [comment]
Merging then

**Review by RealDyllon — APPROVED** [review]

**Pull request merged**